### ISSUE TITLE
KV Cache Quantization

### ISSUE THREAD

**Issue by dinerburger** [EricLBuehler/mistral.rs#971]
Both exllamav2 and llama.cpp support [quantized KV cache](https://github.com/ggerganov/llama.cpp/discussions/5932) to allow pretty large context lengths on consumer hardware. It would be a great addition to mistral.rs; I've been very interested in trying it, but I'm limited to 24GB of VRAM, necessitating, for example, sending KV cache to system RAM instead of keeping it on the card (only possible on llama.cpp to my knowledge).

**Comment by EricLBuehler** [comment]
Hi @dinerburger ! After some recent work in  KV cache, I think we have the infrastructure now for this! I'll take a look again and will probably merge some initial support.

**Comment by dinerburger** [comment]
Obviously there a number of ways to implement KV cache quant, but I'd be interested in knowing which implementation you're considering.

**Comment by EricLBuehler** [comment]
I'm considering two options. The 8 bit cache using FP8 might be easier to implement.
- 4-bit cache: something similar to what exllamav2 does [here](https://github.com/turboderp/exllamav2/commit/324404ebe4e3c4dd0447ffc1290c312de1df02be#diff-144d2aca644ed440b9a057a45d082eb7015da25396352f76b2e84e66f6e0a57b), where we apply a Hadamard transform to reduce the outliers (paper: https://arxiv.org/pdf/2404.00456), and then can use a Q4 cache
- 8bit cache: use FP8, which might be easier & quicker initially
  - We should probably consider using E5M2 rather than E4M3 because of [this](https://github.com/turboderp/exllamav2/blob/4f83f52d7d1bd31bf7e8f52c966600eefbbab80e/exllamav2/exllamav2_ext/cuda/cache.cu#L15-L20).

@sammcj I saw your recent PR merge to ollama supporting KV cache quantization - congrats! What method did you take - did you just quantize the K/V blocks?

**Comment by dinerburger** [comment]
Perfect, yeah I was gonna recommend the Hadamard transform approach. It's easy and effective. I followed that PR pretty closely; @sammcj piggy-backed on llama.cpp's implementation, utilizing either `q4_0` or `q8_0` quant types provided by llama.cpp. Technically llama.cpp is capable of using any of the non-I quant types for KV cache quantization, and these types can be mixed assuming you build llama.cpp yourself with the `GGML_CUDA_FA_ALL_QUANTS` define. Depending on your implementation of the base llama quants this may be appropriate.

**Comment by sammcj** [comment]
Thanks @EricLBuehler! Was simple compared to the efforts you'll be putting in I'm sure as llama.cpp does the heavy lifting of performing the quantisation. The changes to Ollama were mainly around the parameterisation of the Ollama components to make use of it, some memory management for their layer estimations/placement and a lot of shall we say 'soft skills' to get it across the line 😅

You can see the initial changes (bundled with FA support) in llama.cpp here: https://github.com/ggerganov/llama.cpp/pull/7527

While 4bit works well for Exllamav2's KV, the quantisation that works well with llama.cpp/gguf is Q8_0, which is approximately 8.5bpw.

I've published a F16 vs Q8_0 KV perplexity measurement [here](https://smcleod.net/2024/12/bringing-k/v-context-quantisation-to-ollama/#perplexity-measurements) (I might add q4_0 and another dataset variant as well in the next day or two).

Forgive my ignorance here - when you say int4/int8 - are you talking about quantising down to 4/8bit integers, or simply rounding to them?

The reason I ask this is I know that int4/8 models tend to be quite a bit lower quality than their quantised counterparts such as Q4_K_M/Q8_0.

**Comment by EricLBuehler** [comment]
@sammcj @dinerburger sorry for the late reply! I've begun work in #988.

> Forgive my ignorance here - when you say int4/int8 - are you talking about quantising down to 4/8bit integers, or simply rounding to them?

I'll be using Q4_K_M and Q8_0 + a hadamard transform in the kernel for better distributions, not int4/int8.

> Technically llama.cpp is capable of using [EDIT]many of their[/EDIT] quant types for KV cache quantization

Sounds like an interesting idea. I'm curious if we can do something similar after the initial KV cache quantization support is merged.

**Comment by dinerburger** [comment]
Yeah, you can see the supported quant types here: https://github.com/ggerganov/llama.cpp/blob/26a8406ba9198eb6fdd8329fa717555b4f77f05f/common/common.cpp#L1018. A note however, if you want to experiment: compile llama.cpp with `GGML_CUDA_FA_ALL_QUANTS` or else you'll be limited to `Q4_0` and `Q8_0` with no mixing. With that flag, however, you can mix and match different K and V types, which can be nice since in my experience K cache is far more sensitive to quantization than V cache, especially for models leveraging GQA.

**Comment by dinerburger** [comment]
Circling around on this, Transformers allows the use of [HQQ for KV cache quantization](https://huggingface.co/docs/transformers/v4.50.0/en/internal/generation_utils#transformers.HQQQuantizedCache). Since you've already got HQQ integrated, it might be a faster way to integrate KV cache quantization.

**Comment by mahmoodsh36** [comment]
hello,
afaict kv cache quantization is not yet available in mistral-rs? i have been using it with llama-server (llama-cpp) because it allows me to use a considerably longer context size with my 4090 running 32b models. without quantization i couldnt go beyond ~16k context length on this setup.

**Comment by EricLBuehler** [comment]
@sammcj @dinerburger @mahmoodsh36 KV quant is finally implemented in #1400!

**Comment by dinerburger** [comment]
Excellent work, thank you!

**Comment by sammcj** [comment]
Love your work Eric!
